modules/face/src/face_alignment.cpp

```diff
@@ -0,0 +1,296 @@
+            temp.push_back(pt);
+        }
+        trainlandmarks.push_back(temp);
+        f1.close();
+    }
+    return true;
+}
+bool FacemarkKazemiImpl :: getMeanShapeRelative(vector<Rect> face,vector< vector<Point2f> > & initialshape){
+    if(meanshape.empty()) {
+            // throw error if no data (or simply return -1?)
+            String error_message = "The data is not loaded properly by train function. Aborting...";
+            CV_Error(Error::StsBadArg, error_message);
+            return false;
+    }
+    Point2f srcTri[3];
+    //source points to find warp matrix
+    srcTri[0] = Point2f( minmeanx , minmeany );
+    srcTri[1] = Point2f( maxmeanx, minmeany );
+    srcTri[2] = Point2f( minmeanx, maxmeany );
+    Point2f dstTri[3];
+    for(unsigned long k=0;k<face.size();k++){
+        vector<Point2f> temp;
+        //destination points to which the image has to be warped
+        dstTri[0] = Point2f(float(face[k].x) , float(face[k].y) );
+        dstTri[1] = Point2f(float(face[k].x + (1.1)*face[k].width), float(face[k].y) );
+        dstTri[2] = Point2f( float(face[k].x), float(face[k].y+(1.35)*face[k].height));
+        //loop to calculate bounding rectangle of the mean shape found
+        Mat warp_mat( 2, 3, CV_32FC1 );
+        //get affine transform to calculate warp matrix
+        warp_mat = getAffineTransform( srcTri, dstTri );
+        //loop to initialize initial shape
+        for(unsigned long i=0;i<meanshape.size();i++){
+            Point2f pt1=meanshape[i];
+            Mat C = (Mat_<double>(3,1) << pt1.x, pt1.y, 1);
+            Mat D =warp_mat*C;
+            pt1.x=float(abs(D.at<double>(0,0)));
+            pt1.y=float(abs(D.at<double>(1,0)));
+            temp.push_back(pt1);
+        }
+        initialshape.push_back(temp);
+    }
+    return true;
+}
+class doSum : public ParallelLoopBody
+{
+    public:
+        doSum(vector<training_sample>* samples_,vector<Point2f>* sum_) :
+        samples(samples_),
+        sum(sum_)
+        {
+        }
+        virtual void operator()( const cv::Range& range) const
+        {
+            for (size_t j = range.start; j <(size_t) range.end; ++j){
+                for(unsigned long k=0;k<(*samples)[j].shapeResiduals.size();k++){
+                    (*sum)[k]=(*sum)[k]+(*samples)[j].shapeResiduals[k];
+                }
+            }
+        }
+    private:
+        vector<training_sample>* samples;
+        vector<Point2f>* sum;
+};
+bool FacemarkKazemiImpl :: calcMeanShape (vector< vector<Point2f> >& trainlandmarks,InputArrayOfArrays _trainimages){
+    //get vector of images
+    vector<Mat> trainimages;
+    //clear the loaded meanshape
+    meanshape.clear();
+    _trainimages.getMatVector(trainimages);
+    if(trainimages.empty()||trainlandmarks.size()!=trainimages.size()) {
+        // throw error if no data (or simply return -1?)
+        String error_message = "The data is not loaded properly by train function. Aborting...";
```

Comment:
could you somehow return why? because this is kind of a large range of possibilities.